### PR TITLE
Updated proofs harnesses for aws_cryptosdk_rsa_encrypt and decrypt

### PR DESCRIPTION
**Pull request by tegansb**
*Description of changes:*

Updated proof harnesses for aws_cryptosdk_rsa_encrypt and aws_cryptosdk_rsa_decryrpt to work with may-fail-malloc flag. 
Added preconditions to source code. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: .cbmc-batch/jobs/aws_cryptosdk_rsa_decrypt/aws_cryptosdk_rsa_decrypt_harness.c [hunk 1]
```diff
@@ -21,31 +21,38 @@
...
 void aws_cryptosdk_rsa_decrypt_harness() {
+    /* Nondet Inputs */
     struct aws_byte_buf plain;
     struct aws_allocator *alloc = can_fail_allocator();
     struct aws_byte_cursor cipher;
     struct aws_string *key = ensure_string_is_allocated_bounded_length(KEY_LEN);
     enum aws_cryptosdk_rsa_padding_mode rsa_padding_mode;
 
+    /* Assumptions */
+    __CPROVER_assume(key != NULL);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_rsa_decrypt/aws_cryptosdk_rsa_decrypt_harness.c:32]: Why can't we assume here `aws_string_is_valid()`?

File: .cbmc-batch/jobs/aws_cryptosdk_rsa_encrypt/aws_cryptosdk_rsa_encrypt_harness.c [hunk 2]
```diff
@@ -21,31 +21,38 @@
...
 void aws_cryptosdk_rsa_encrypt_harness() {
+    /* Nondet Inputs */
     struct aws_byte_buf cipher;
     struct aws_allocator *alloc = can_fail_allocator();
     struct aws_byte_cursor plain;
     struct aws_string *key = ensure_string_is_allocated_bounded_length(KEY_LEN);
     enum aws_cryptosdk_rsa_padding_mode rsa_padding_mode;
 
+    /* Assumptions */
+    __CPROVER_assume(key != NULL);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_rsa_encrypt/aws_cryptosdk_rsa_encrypt_harness.c:32]: Why can't we assume here `aws_string_is_valid()`?

File: source/cipher.c [hunk 3]
```diff
@@ -928,6 +928,9 @@ int aws_cryptosdk_rsa_encrypt(
     const struct aws_byte_cursor plain,
     const struct aws_string *rsa_public_key_pem,
     enum aws_cryptosdk_rsa_padding_mode rsa_padding_mode) {
+    AWS_PRECONDITION(aws_byte_buf_is_valid(cipher));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&plain));
+    AWS_PRECONDITION(rsa_public_key_pem != NULL);
```
Comment (feliperodri) [inline source/cipher.c:933]: Why can't we assume here `aws_string_is_valid()`?

File: source/cipher.c [hunk 4]
```diff
@@ -976,6 +979,9 @@ int aws_cryptosdk_rsa_decrypt(
     const struct aws_byte_cursor cipher,
     const struct aws_string *rsa_private_key_pem,
     enum aws_cryptosdk_rsa_padding_mode rsa_padding_mode) {
+    AWS_PRECONDITION(aws_byte_buf_is_valid(plain));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&cipher));
+    AWS_PRECONDITION(rsa_private_key_pem != NULL);
```
Comment (feliperodri) [inline source/cipher.c:984]: Why can't we assume here `aws_string_is_valid()`?

**Review by feliperodri — CHANGES_REQUESTED** [review]

File: .cbmc-batch/jobs/aws_cryptosdk_rsa_decrypt/aws_cryptosdk_rsa_decrypt_harness.c (hunk 1 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_rsa_decrypt/aws_cryptosdk_rsa_decrypt_harness.c:32]: Updated and also using the new c-common functions.

File: .cbmc-batch/jobs/aws_cryptosdk_rsa_encrypt/aws_cryptosdk_rsa_encrypt_harness.c (hunk 2 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_rsa_encrypt/aws_cryptosdk_rsa_encrypt_harness.c:32]: Updated!

File: source/cipher.c (hunk 3 above)
Comment (tegansb) [reply source/cipher.c:933]: Updated!

File: source/cipher.c (hunk 4 above)
Comment (tegansb) [reply source/cipher.c:984]: Updated!

File: .cbmc-batch/jobs/aws_cryptosdk_rsa_encrypt/aws_cryptosdk_rsa_encrypt_harness.c [hunk 5]
```diff
@@ -21,31 +21,39 @@
...
     struct aws_byte_buf cipher;
     struct aws_allocator *alloc = can_fail_allocator();
     struct aws_byte_cursor plain;
-    struct aws_string *key = ensure_string_is_allocated_bounded_length(KEY_LEN);
+    struct aws_string *key = ensure_string_is_allocated_nondet_length();
     enum aws_cryptosdk_rsa_padding_mode rsa_padding_mode;
 
+    /* Assumptions */
+    __CPROVER_assume(aws_string_is_valid(key));
+    __CPROVER_assume(key->len <= KEY_LEN);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_rsa_encrypt/aws_cryptosdk_rsa_encrypt_harness.c:33]: We should have a "bounded" allocated nondet length predicate, so we don't have to assume the bound here (but this is not a blocker for this PR).

**Review by feliperodri — APPROVED** [review]
LGTM.

File: .cbmc-batch/jobs/aws_cryptosdk_rsa_encrypt/aws_cryptosdk_rsa_encrypt_harness.c (hunk 5 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_rsa_encrypt/aws_cryptosdk_rsa_encrypt_harness.c:33]: Yeah I've had to do this in other PRs to so let me make a github issue about it in c-common.

**Pull request merged**